### PR TITLE
Fix desktop lifetime non-mainwindow cancellation

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

That's a rather old bug which was surfaced after my previous fix: https://github.com/AvaloniaUI/Avalonia/pull/15438
Current PR makes Avalonia behave closer to WPF:
When ShutdownMode is set to OnMainWindowClose, only main window can cancel shutdown.

## What is the current behavior?

Any window can cancel app shutdown, even after main window was closed.

## What is the updated/expected behavior with this PR?

Only main window shutdown cancellation is respected, when ShutdownMode=OnMainWindowClose

## Checklist

- [x] Added unit tests (if possible)?

## Breaking changes

Possible behavioral breaking change. At the same time, new behavior might be an expected one for WPF developers.

### CODE REVIEW

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs [hunk 1]
```diff
@@ -184,19 +183,24 @@ private bool DoShutdown(
                 // When an OS shutdown request is received, try to close all non-owned windows. Windows can cancel
                 // shutdown by setting e.Cancel = true in the Closing event. Owned windows will be shutdown by their
                 // owners.
-                foreach (var w in Windows.ToArray())
+                foreach (var w in new List<Window>(_windows))
```
Comment (MrJul) [inline src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:186]: Nit: why the extra `List`? Not that the extra single allocation will matter in this case, but it also seems unnecessary compared to the array used previously.

**Review by MrJul — COMMENTED** [review]
What do you think about ignoring `WindowClosingEventArgs.Cancel` in `Window.CloseCore` on shutdown (and document it) for the secondary windows, effectively closing them (and matching WPF)?

Currently, the extra windows won't ever have a chance of running any cleanup code.

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:186]: Just a minor change to remove Linq usage. As we need a copy of the windows collection.

**Comment by maxkatz6** [comment]
> What do you think about ignoring WindowClosingEventArgs.Cancel in Window.CloseCore on shutdown (and document it) for the secondary windows, effectively closing them (and matching WPF)?

> Currently, the extra windows won't ever have a chance of running any cleanup code.

But that's how it works in this PR right now.
I.e. secondary window also gets `Closing` event raised, but Cancel value is ignored.
Which is actually different from WPF - only MainWindow gets this event raised.

Sorry, somehow I missed these review comments before.

**Comment by MrJul** [comment]
> But that's how it works in this PR right now.
> I.e. secondary window also gets `Closing` event raised, but Cancel value is ignored.
> Which is actually different from WPF - only MainWindow gets this event raised there.

Sorry, I wasn't very clear.
It's `Closed` that isn't being raised for the secondary windows with this PR, preventing them from running any save/cleanup code they might need.

(My initial comment mentioned ignoring `Cancel` in `CloseCore` which I meant as "we should still close the window".)

AFAICT, in WPF, both `Closing` and `Closed` are raised for _all_ windows when `ShutdownMode == OnMainWindowClose`.

**Comment by maxkatz6** [comment]
@MrJul Window.Closed event is raised properly now.
I also changed behavior of `Lifetime.Shutdown()` method to match the same rules - execute Closed even though Closing was cancelled.

**Review by MrJul — APPROVED** [review]

**Comment by MrJul** [comment]
> @MrJul Window.Closed event is raised properly now.
> I also changed behavior of `Lifetime.Shutdown()` method to match the same rules - execute Closed even though Closing was cancelled.

Nice! As a bonus I find the new code to be easier to follow :)

**Pull request merged**